tests/test_backend_failure.py

```diff
@@ -0,0 +1,219 @@
+@with_path(paths=['/1.root.json', '/root.json', '/targets.json'])
+@with_director(handlers=[
+                            DownloadInterruptionHandler(number_of_failures=1),
+                            MalformedJsonHandler(number_of_failures=1),
+                            DownloadInterruptionHandler(number_of_failures=4),
+                            MalformedJsonHandler(number_of_failures=4)
+                        ])
+@with_secondary()
+@with_aktualizr(run_mode='full')
+@with_install_manager()
+def test_backend_failure_sanity_director_update_after_metadata_download_failure(install_mngr, director, **kwargs):
+    director.wait_for_install()
+    return director.is_install_successful() and install_mngr.are_images_installed()
+
+
+"""
+Verifies whether aktualizr is updatable after image metadata download failure
+with follow-up successful metadata download.
+
+Currently, it's tested against two types of metadata download/parsing failure:
+    - download interruption - metadata file download is interrupted once|four times, after that it's successful
+    - malformed json - metadata file download is successful but the received metadata are malformed.
+      It happens at once| four time, after that it's successful
+
+
+Note: Aktualizr doesn't send any installation report in manifest in case of metadata download failure
+"""
+@with_uptane_backend(start_generic_server=True, port=8888)
+@with_path(paths=['/1.root.json', '/root.json', '/timestamp.json', '/snapshot.json', '/targets.json'])
+@with_director()
+@with_imagerepo(handlers=[
+                            DownloadInterruptionHandler(number_of_failures=1),
+                            MalformedJsonHandler(number_of_failures=1),
+                            DownloadInterruptionHandler(number_of_failures=4),
+                            MalformedJsonHandler(number_of_failures=4)
+                        ])
+@with_secondary()
+@with_aktualizr(run_mode='full')
+@with_install_manager()
+def test_backend_failure_sanity_imagerepo_update_after_metadata_download_failure(install_mngr, director, **kwargs):
+    director.wait_for_install()
+    return director.is_install_successful() and install_mngr.are_images_installed()
+
+
+"""
+Verifies whether aktualizr is updatable after image download failure
+with follow-up successful download.
+
+Currently, it's tested against two types of image download failure:
+    - download interruption - file download is interrupted once, after that it's successful
+    - malformed image - image download is successful but it's malformed. It happens at once after that it's successful
+"""
+@with_uptane_backend(start_generic_server=True, port=8888)
+@with_images(images_to_install=[(('primary-hw-ID-001', 'primary-ecu-id'), 'primary-image.img')])
+@with_director()
+@with_imagerepo(handlers=[
+                            # TODO: test fails because aktualizr issues byte range request
+                            # that are not supported by server
```

Comment:
By "not supported by server", you mean the test server being used her, right? I was under the impression the real server handled this, and this is already covered by the `fetcher_test` and `fetcher_death_test` as well, unless those tests don't do what I think they do.

Reply:
> By "not supported by server", you mean the test server being used her, right? I was under the impression the real server handled this, and this is already covered by the `fetcher_test` and `fetcher_death_test` as well, unless those tests don't do what I think they do.

Yes, I mean the test server being used  in the given test suite. That's correct, the current real/OTA Connect image repo does support a byte range request, in case of the `custom URL` usage we don't know in advance if the `custom image server` supports it, so IMHO, aktualizr needs to be able to fallback to regular HTTP GET requests during download retries as long as we need to support `custom URL`.

Reply:
> this is already covered by the fetcher_test and fetcher_death_test as well, unless those tests don't do what I think they do.
`fetcher_test` doesn't test/simulate the given case, it uses the mocked implementation of http client which ignores a range/offset value.

`fetcher_death_test` tests similar use case (resume after pause/die) but it tests it against a server that does support byte range requests (fake_test_server.py::elif self.path.endswith('/large_file'): )

Reply:
> aktualizr needs to be able to fallback to regular HTTP GET requests during download retries

I understand and agree. Actually, it sounds like we need to be able to fall back to regular GET requests regardless of the custom URL usage. We should also try to see if we can get byte range requests working in treehub so that we can take advantage of the functionality we already have, too.